test/functional/autocmd/tabnewentered_spec.lua

```diff
@@ -386,18 +399,32 @@ describe('tabpage/previous', function()
     switches_to_previous_after_switching_to_previous('g<Tab>', '<C-W>g<Tab>'))
   it('switches to previous via <C-Tab> after switching to previous via g<Tab>',
     switches_to_previous_after_switching_to_previous('g<Tab>', '<C-Tab>'))
+  it('switches to previous via :tabn #<CR> after switching to previous via g<Tab>',
+    switches_to_previous_after_switching_to_previous('g<Tab>', ':tabn #<CR>'))
   it('switches to previous via g<Tab> after switching to previous via <C-W>g<Tab>',
     switches_to_previous_after_switching_to_previous('<C-W>g<Tab>', 'g<Tab>'))
   it('switches to previous via <C-W>g<Tab> after switching to previous via <C-W>g<Tab>',
     switches_to_previous_after_switching_to_previous('<C-W>g<Tab>', '<C-W>g<Tab>'))
   it('switches to previous via <C-Tab> after switching to previous via <C-W>g<Tab>',
     switches_to_previous_after_switching_to_previous('<C-W>g<Tab>', '<C-Tab>'))
+  it('switches to previous via :tabn #<CR> after switching to previous via <C-W>g<Tab>',
+    switches_to_previous_after_switching_to_previous('<C-W>g<Tab>', ':tabn #<CR>'))
   it('switches to previous via g<Tab> after switching to previous via <C-Tab>',
     switches_to_previous_after_switching_to_previous('<C-Tab>', 'g<Tab>'))
   it('switches to previous via <C-W>g<Tab> after switching to previous via <C-Tab>',
     switches_to_previous_after_switching_to_previous('<C-Tab>', '<C-W>g<Tab>'))
   it('switches to previous via <C-Tab> after switching to previous via <C-Tab>',
     switches_to_previous_after_switching_to_previous('<C-Tab>', '<C-Tab>'))
+  it('switches to previous via :tabn #<CR> after switching to previous via <C-Tab>',
+    switches_to_previous_after_switching_to_previous('<C-Tab>', ':tabn #<CR>'))
+  it('switches to previous via g<Tab> after switching to previous via :tabn #<CR>',
+    switches_to_previous_after_switching_to_previous(':tabn #<CR>', 'g<Tab>'))
+  it('switches to previous via <C-W>g<Tab> after switching to previous via :tabn #<CR>',
+    switches_to_previous_after_switching_to_previous(':tabn #<CR>', '<C-W>g<Tab>'))
+  it('switches to previous via <C-Tab> after switching to previous via <C-Tab>',
+    switches_to_previous_after_switching_to_previous(':tabn #<CR>', '<C-Tab>'))
+  it('switches to previous via :tabn #<CR> after switching to previous via :tabn #<CR>',
+    switches_to_previous_after_switching_to_previous(':tabn #<CR>', ':tabn #<CR>'))
```

Comment:
Your thorough tests are very appreciated :) 

General note for the future: there is a point of too many tests--behavior can be over-specified, then it's hard to change later. E.g. in this case, testing `'<C-Tab>', ':tabn #<CR>'` is good enough, the other variations aren't really testing anything different, and that codepath is very unlikely to change. 

Tests are a tradeoff between quality vs. CI cost (time) + over-specification (hard to change behavior because many tests must be changed).

Reply:
Thank you for the feedback!  Those are great points, and I will keep that in mind for future patches.